File: Leetcode/Reverse Words in a String III.cpp
class Solution {
    void reverseWord(int startIndexInclusive, int endIndexInclusive, string &sentence) {
        for (;startIndexInclusive <= endIndexInclusive;++startIndexInclusive, --endIndexInclusive) {
            char temp = sentence[startIndexInclusive];
            sentence[startIndexInclusive] = sentence[endIndexInclusive];
            sentence[endIndexInclusive] = temp;
        }
    }
public:
    string reverseWords(string sentence) {
        for (int lastSpaceIndex = -1, currentIndex = 0; currentIndex <= sentence.size(); ++currentIndex) {
            if (currentIndex == sentence.size() || sentence[currentIndex] == ' ') {
                reverseWord(lastSpaceIndex + 1, currentIndex - 1, sentence);
                lastSpaceIndex = currentIndex;
            }
        }
        return sentence;
    }
};

File: Leetcode/Reverse Words in a String III.java
class Solution {
    private void copyWordInReverse(int startIndexInclusive, int endIndexExclusive,
                                   String s, StringBuilder reversedWordSentence) {
        for (int currentIndex = endIndexExclusive + 1;
             endIndexExclusive < startIndexInclusive;
             ++currentIndex, --startIndexInclusive
        ) {
            reversedWordSentence.setCharAt(currentIndex, s.charAt(startIndexInclusive));
        }
    }

    public String reverseWords(String s) {
        StringBuilder reversedWordSentence = new StringBuilder(s);
        for (int lastSpaceIndex = -1, currentIndex = 0; currentIndex <= s.length(); ++currentIndex) {
            if (currentIndex == s.length() || s.charAt(currentIndex) == ' ') {
                copyWordInReverse(currentIndex - 1, lastSpaceIndex, s, reversedWordSentence);
                if (currentIndex < s.length()) {
                    reversedWordSentence.setCharAt(currentIndex, ' ');
                    lastSpaceIndex = currentIndex;
                }
            }
        }
        return reversedWordSentence.toString();
    }
}
